### PR TITLE
【newir】Concat optest

### PR DESCRIPTION
**Pull request by xiaoguoguo626807**
### PR types

Bug fixes
### PR changes

others
### Description

pcard-67164
修复新ir concat kernel选择问题，data -> combine -> concat 会导致concat 无法感知data op 的place, 选择默认place 导致cpu tensor选择gpu kernel.
修复optest loss.backward 分支grad 接口传入inputs 未过滤问题

### CODE REVIEW

File: python/paddle/tensor/manipulation.py [hunk 1]
```diff
@@ -1131,11 +1132,11 @@ def concat(x, axis=0, name=None):
         if not isinstance(input, Variable):
             input = [t for t in input if t.shape.count(0) == 0]
         return _C_ops.concat(input, axis)
+    elif in_new_ir_mode():
```
Comment (0x45f) [inline python/paddle/tensor/manipulation.py:1135]: in_new_ir_mode分支是不是能和上面in_dynamic_mode分支统一？

File: python/paddle/tensor/math.py [hunk 2]
```diff
@@ -264,6 +267,10 @@ def scale(x, scale=1.0, bias=0.0, bias_after_scale=True, act=None, name=None):
             return _C_ops.scale(x, scale, float(bias), bias_after_scale)
         out = _C_ops.scale(x, scale, float(bias), bias_after_scale)
         return dygraph_utils._append_activation_in_dygraph(out, act)
+    elif in_new_ir_mode():
+        if act is None:
+            return _C_ops.scale(x, scale, float(bias), bias_after_scale)
+        raise ValueError("act is not implement in new ir of scale api.")
```
Comment (0x45f) [inline python/paddle/tensor/math.py:273]: 老静态图分支在act 非 None的情况下也会raise error吗？

File: paddle/fluid/ir/transforms/pd_op_to_kernel_pass.cc [hunk 3]
```diff
@@ -569,6 +569,29 @@ phi::KernelKey GetKernelKey(
...
+                                  .data();
+
+            auto data_op_backend =
+                paddle::experimental::ParseBackend(data_place);
+            if (data_op_backend == phi::Backend::UNDEFINED) {
+              data_op_backend = paddle::experimental::ParseBackend(place);
+            }
+            kernel_key_parser.key_set.backend_set =
+                kernel_key_parser.key_set.backend_set |
+                paddle::experimental::BackendSet(data_op_backend);
```
Comment (YuanRisheng) [inline paddle/fluid/ir/transforms/pd_op_to_kernel_pass.cc:591]: 把这段逻辑封装一下吧，可以放在函数中，比如UpdateBackendOfDataOp

File: paddle/fluid/ir/transforms/pd_op_to_kernel_pass.cc [hunk 4]
```diff
@@ -569,6 +569,29 @@ phi::KernelKey GetKernelKey(
         kernel_key_parser.key_set.backend_set =
             kernel_key_parser.key_set.backend_set |
             paddle::experimental::BackendSet(data_op_backend);
+      } else if (op->operand_source(i).GetDefiningOp()->name() ==
+                 "builtin.combine") {
+        auto combine_op = op->operand_source(i).GetDefiningOp();
```
Comment (zhangbo9674) [inline paddle/fluid/ir/transforms/pd_op_to_kernel_pass.cc:574]: 建议添加对 combine_op 的所有 operands 的 place 进行判等检验

File: python/paddle/tensor/manipulation.py (hunk 1 above)
Comment (xiaoguoguo626807) [reply python/paddle/tensor/manipulation.py:1135]: 此处下一个pr修改，目前concat 破test还有其他问题，解决后续问题时统一此分支

File: python/paddle/tensor/math.py (hunk 2 above)
Comment (xiaoguoguo626807) [reply python/paddle/tensor/math.py:273]: 由于暂时未支持全部act newir api  ,需要分开处理。实现对应的_append_activation_in_newir函数;
老静态图支持helper append_op

File: paddle/fluid/ir/transforms/pd_op_to_kernel_pass.cc (hunk 3 above)
Comment (xiaoguoguo626807) [reply paddle/fluid/ir/transforms/pd_op_to_kernel_pass.cc:591]: 下个pr 修复

File: paddle/fluid/ir/transforms/pd_op_to_kernel_pass.cc (hunk 4 above)
Comment (xiaoguoguo626807) [reply paddle/fluid/ir/transforms/pd_op_to_kernel_pass.cc:574]: 下个pr 修复

**Review by YuanRisheng — APPROVED** [review]

**Review by Aurelius84 — APPROVED** [review]
